CPF-2218/SecureDAOs

jlhughes opened the pull request:
Addressing Penetration Testing results which revealed a number of improperly authenticated DAOs. 

All DAOs need to be migrated to EasyDAO, but this first set are significant.

jlhughes committed:
Add support for:
* setting index properties - which are added to MDAO before JDAO setup
* setting orderBy properties
* providing a delegate DAO which is inserted into the delegate chain.

jlhughes committed:
Call planSelect with state array rather than the first state as it is expecting an array.

jlhughes committed:
Refactor to work with EasyDAO

* move delegate.select to setDelegate
* let EasyDAO or caller setup delegate as JDAO.

jlhughes committed:
Do not serve DAOs with NSpec.serve = false

jlhughes committed:
Reconfigure a number of DAOs serviceScript to use EasyDAO so Authenticated properly.

jlhughes committed:
Correct dugDAO journal name from digs to dugs.

kgrgreer commented on src/foam/dao/EasyDAO.js:
```diff
@@ -102,20 +107,41 @@ foam.CLASS({
   setMdao((foam.dao.MDAO)delegate);
+  if ( getIndex() != null &&
+       getIndex().length > 0 ) {
+    // IMPORTANT: Indexes must be added before JDAO replay occurs.
```
I don't think this is true, but it doesn't hurt.

jlhughes replied:
At the moment this is the case, DAOs are simply broken otherwise.

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-2218/SecureDAOs

jlhughes committed:
Correct log message reporting where journal is being created.

jlhughes committed:
Change order to array of Comparator rather than PropertyInfo

Also, support a chain of ProxyDAOs as the decorator. Walk that chain and set the delegate on the last in list.

jlhughes committed:
Split logMessageDAO into unauthenticated localLogMessageDAO and authenticated logMessageDAO.
Update DAOLogger to now look for context key 'localLostMessageDAO'.

Also create an authenticated sessionDAO.

jlhughes committed:
Split emailMessageDAO into unauthenticated localEmailMessageDAO and authenticated emailMessageDAO.
Update emailMessageDAO context key to localEmailMessageDAO

jlhughes committed:
Respond with error when requested service not found.

kgrgreer commented on src/foam/dao/EasyDAO.js:
```diff
@@ -388,6 +426,16 @@ return delegate;
+    {
+      class: 'FObjectArray',
+      of: 'foam.core.PropertyInfo',
+      name: 'index'
```
Just adding on index isn't very useful. Better if this took an array of indices (each of which is an array of PropertyInfos.

jlhughes replied:
@kgrgreer

kgrgreer commented on src/foam/dao/EasyDAO.js:
```diff
@@ -189,6 +222,11 @@ return delegate;
+    {
+      class: 'Object',
+      type: 'foam.dao.DAO',
+      name: 'decorator'
```
This would be more useful if there were more than one.

jlhughes replied:
@kgrgreer It handles a ProxyDAO - walks the chain and sets the delegate on the last.

kgrgreer commented on src/foam/dao/EasyDAO.js:
```diff
@@ -162,6 +188,13 @@ if ( getContextualize() ) {
+if ( getOrder() != null &&
+     getOrder().length > 0 ) {
+  for ( foam.core.PropertyInfo prop : getOrder() ) {
+    delegate = delegate.orderBy(prop);
```
This will just order by the last property. You want a CompoundComparator, which we currently only support in JS, but we should fix/implement that. Will also require work on MDAO to support.

jlhughes replied:
Presently serviceScripts only have one orderBy so this will work for now and we can add support in another PR.

kgrgreer commented on src/foam/dao/EasyDAO.js:
```diff
@@ -162,6 +188,13 @@ if ( getContextualize() ) {
 }
 
+if ( getOrder() != null &&
+     getOrder().length > 0 ) {
```
The 'order' property isn't an array, so .length doesn't make sense.

jlhughes replied:
@kgrgreer 'order' is an FObjectArray of foam.mlang.order.Comparator

TW80000 replied:
@jlhughes Wouldn't we just end up ordering by whatever the last Comparator in the array was? If so, it doesn't make sense for 'order' to be an Array.

jlhughes replied:
'order' is an array, the issue is the dao.orderBy call which needs to compose them in some way.

kgrgreer commented on src/foam/dao/EasyDAO.js:
```diff
@@ -576,6 +624,12 @@ return delegate;
         dao = decorated;
       }
 
+      if ( this.order ) {
```
Didn't you already do this above? Also, order isn't an array.

jlhughes replied:
@kgrgreer one is java, this is javascript.

kgrgreer commented on src/foam/dao/index/AltIndex.java:
```diff
@@ -37,13 +37,13 @@ public Object addIndex(Object state, Index i) {
     };
 
-    planSelect(sa[0], sink, 0, Long.MAX_VALUE, null, null).select(sa[0], sink, 0, Long.MAX_VALUE, null, null);
+    planSelect(sa, sink, 0, Long.MAX_VALUE, null, null).select(sa[0], sink, 0, Long.MAX_VALUE, null, null);
```
Should the second reference to sa[0] also be converted to 'sa'?

jlhughes replied:
@kgrgreer it's actually ignored. So more correctly would be to pass null.

jlhughes replied:
@kgrgreer planSelect returns an AltSelectPlan which uses it's internal state_ rather than the argument state passed on the select call.

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-2218/SecureDAOs

jlhughes committed:
Clean up log messages and use NSpec name when available for EasyDAO name.

jlhughes committed:
Log error through Logger when available.

jlhughes committed:
Log through System.err when Logger is not available.

jlhughes committed:
Correct syntax error getOwnclassInfo -> getOwnClassInfo

jlhughes committed:
Correct testDAO journal file name: test -> tests
Correct permissionDAO setAuthenticated -> setAuthenticate

jlhughes committed:
Update Prevent Priviledge test to test adding permission when one has the permission.

jlhughes committed:
Correct syntax error - catch t but print stack trace of e

jlhughes committed:
Add context only constructor so can be constructed by EasyDAO - before context is set.

jlhughes committed:
Split permissionDAO and groupPermissionJunctionDAO into local and non-local to restict access from client.
Replace context lookups of non-local to local

TW80000 commented on src/foam/dao/AuthenticatedDAO.java:
```diff
@@ -61,7 +61,7 @@ public FObject put_(X x, FObject obj) {
       permission = createPermission("update", id);
     }
-
+    ((foam.nanos.logger.Logger)x.get("logger")).debug("check put permission", permission);
```
Should probably be removed before merging.

TW80000 commented on src/foam/nanos/auth/Relationships.js:
```diff
@@ -10,7 +10,7 @@ foam.RELATIONSHIP({
   forwardName: 'permissions',
   inverseName: 'groups',
-  junctionDAOKey: 'groupPermissionJunctionDAO'
+  junctionDAOKey: 'localGroupPermissionJunctionDAO'
```
I don't think we want to do this. If we do, I think client-side code related to relationships will try to use `localGroupPermissionJunctionDAO` from the client and fail since it isn't served.

jlhughes replied:
@TW80000 why would client side code be accessing this dao?

TW80000 replied:
If you go to the DetailView for either of those models (Group and Permission) and try to edit the relationship using the GUI, it's going to try to read from and write to the junctionDAO, which won't be served if you change this to the local version. It would also break the permission matrix, which needs to read from groupPermissionJunctionDAO on the client to know which permissions the different groups have.

TW80000 commented on src/services:
```diff
@@ -82,14 +83,17 @@ p({"class":"foam.nanos.boot.NSpec", "name":"supportEmailDAO",                  "
 p({"class":"foam.nanos.boot.NSpec", "name":"serializationTestEchoService", "boxClass": "foam.nanos.test.EchoServiceSkeleton", "lazy":true, "serve":true, "authenticate": true, "serviceClass":"foam.nanos.test.SerializationTestEchoService", "client":"{\"class\":\"foam.nanos.test.ClientEchoService\",\"delegate\":{\"class\":\"foam.box.SessionClientBox\",\"delegate\":{\"class\":\"foam.box.HTTPBox\",\"m ...
 p({"class":"foam.nanos.boot.NSpec","name":"themeDAO","description":"Stores Themes, which are used to style the application.","serve":true,"authenticate":false,"serviceScript":"return new foam.dao.EasyDAO.Builder(x)\u000a  .setAuthenticateRead(false)\u000a  .setAuthenticate(true)\u000a  .setPm(true)\u000a  .setSeqNo(true)\u000a  .setOf(foam.nanos.theme.Theme.getOwnClassInfo())\u000a  .setJournalTy ...
-p({"class":"foam.nanos.boot.NSpec","name":"groupPermissionJunctionDAO","description":"The junction DAO for the many-to-many relationship between groups and permissions","serve":true,"serviceScript":"mdao = new foam.dao.MDAO(foam.nanos.auth.GroupPermissionJunction.getOwnClassInfo()); ed = new foam.dao.EasyDAO.Builder(x)\u000a  .setOf(foam.nanos.auth.GroupPermissionJunction.getOwnClassInfo())\u000a ...
+//p({"class":"foam.nanos.boot.NSpec","name":"groupPermissionJunctionDAO","description":"The junction DAO for the many-to-many relationship between groups and permissions","serve":true,"serviceScript":"mdao = new foam.dao.MDAO(foam.nanos.auth.GroupPermissionJunction.getOwnClassInfo()); ed = new foam.dao.EasyDAO.Builder(x)\u000a  .setOf(foam.nanos.auth.GroupPermissionJunction.getOwnClassInfo())\u00 ...
```
We shouldn't commit commented out code if we don't need it.

TW80000 reviewed (commented):
I'm getting this error when I check out this branch and try to use the client:

```
2019.07.09.13.52.24.566,DEBUG,1348 - admin,[Service],nSpecDAO,returning skeleton exception,java.lang.NullPointerException
	at foam.nanos.auth.CachingAuthService.getPermissionMap(CachingAuthService.java:82)
	at foam.nanos.auth.CachingAuthService.check(CachingAuthService.java:108)
	at foam.nanos.auth.ProxyAuthService.check(ProxyAuthService.java:241)
	at foam.nanos.auth.EnabledCheckAuthService.check(EnabledCheckAuthService.java:42)
	at net.nanopay.meter.compliance.ComplianceAuthService.check(ComplianceAuthService.java:71)
	at foam.nanos.auth.ProxyAuthService.check(ProxyAuthService.java:241)
	at foam.nanos.auth.ProxyAuthService.check(ProxyAuthService.java:241)
	at foam.nanos.auth.ProxyAuthService.check(ProxyAuthService.java:241)
	at foam.nanos.auth.ProxyAuthService.check(ProxyAuthService.java:241)
	at foam.nanos.auth.twofactor.TwoFactorAuthService.check(TwoFactorAuthService.java:31)
	at foam.nanos.auth.PermissionedPropertyDAO.maybeRemoveProperties(PermissionedPropertyDAO.java:90)
	at foam.nanos.auth.PermissionedPropertyDAO.find_(PermissionedPropertyDAO.java:32)
	at foam.dao.ProxyDAO.find_(ProxyDAO.java:235)
	at foam.dao.AuthenticatedDAO.find_(AuthenticatedDAO.java:95)
	at foam.dao.ProxyDAO.find_(ProxyDAO.java:235)
	at foam.dao.DAOSkeleton.send(DAOSkeleton.java:147)
	at foam.dao.SessionDAOSkeleton.send(SessionDAOSkeleton.java:51)
	at foam.box.SessionServerBox.send(SessionServerBox.java:153)
	at foam.nanos.http.ServiceWebAgent.execute(ServiceWebAgent.java:118)
	at foam.nanos.http.NanoRouter.service(NanoRouter.java:85)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:865)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1655)
	at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:215)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
	at net.nanopay.security.csp.CSPFilter.doFilter(CSPFilter.java:24)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:533)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1253)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1155)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:531)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:352)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:760)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:678)
	at java.lang.Thread.run(Thread.java:748)
```

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-2218/SecureDAOs

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-2218/SecureDAOs

jlhughes committed:
Remove accidentally committed debug statement

jlhughes committed:
Add permissionPrefix and readOnly

- permissionPrefix: override default AuthenticatedDAO.name taken from 'js: of.name' or 'java: of.objClass.simpleName'

jlhughes committed:
Add constructors compatible with EasyDAO.decorator setup

jlhughes committed:
Remove previous commented out groupPermissionJunctionDAO setup

jlhughes committed:
Adjust tests to allow re-running without clearing journals.

jlhughes committed:
Revert target dao key to non-local groupPermissionJunctionDAO

jlhughes committed:
Correct serviceScript for groupPermissionJunctionDAO it was missing it's inner DAO

TW80000 commented on src/services:
```diff
@@ -90,8 +90,9 @@ p({"class":"foam.nanos.boot.NSpec", "name":"ruleHistoryDAO",                   "
+
 p({"class":"foam.nanos.boot.NSpec","name":"localGroupPermissionJunctionDAO","description":"The junction DAO for the many-to-many relationship between groups and permissions","serve":false,"serviceScript":"return new foam.dao.EasyDAO.Builder(x).setOf(foam.nanos.auth.GroupPermissionJunction.getOwnClassInfo()).setAuthenticate(false).setJournalType(foam.dao.JournalType.SINGLE_JOURNAL).setJournalName( ...
-p({"class":"foam.nanos.boot.NSpec","name":"groupPermissionJunctionDAO","description":"The junction DAO for the many-to-many relationship between groups and permissions","serve":true,"serviceScript":"return new foam.dao.EasyDAO.Builder(x).setOf(foam.nanos.auth.GroupPermissionJunction.getOwnClassInfo()).setInnerDAO((foam.dao.DAO) x.get(\"\")).build();","client":"{\"of\": \"foam.nanos.auth.GroupPerm ...
+p({"class":"foam.nanos.boot.NSpec","name":"groupPermissionJunctionDAO","description":"The junction DAO for the many-to-many relationship between groups and permissions","serve":true,"serviceScript":"return new foam.dao.EasyDAO.Builder(x).setOf(foam.nanos.auth.GroupPermissionJunction.getOwnClassInfo()).setAuthenticate(true).setInnerDAO((foam.dao.DAO) x.get(\"localGroupPermissionJunctionDAO\")).bui ...
```
We don't need `setAuthenticate(true)`. The `GroupPermissionJunction` model implements the `Authorizable` interface. `groupPermissionJunctionDAO` should be wrapped in an `AuthorizationDAO`.

jlhughes committed:
Add support for AuthorizationDAO

jlhughes committed:
Setting permission and groupPermissionJunction services as I think they should be for continue troubleshooting/testing.

jlhughes committed:
REvert groupPermissionJunctionDAO to nom-local

jlhughes committed:
Handle null NSpec during test case runs

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-2218/SecureDAOs

jlhughes committed:
Correct AuthorizationDAO setup, instanceof test was incorrectly getOf rather than getOf.getObjClass.

jlhughes committed:
Add Authorization to localGroupPermissionJunctionDAO
Revert all uses of groupPermissionJunctionDAO to non-local.

jlhughes committed:
On authorizeOnUpdate handle null parent group

TW80000 commented on src/foam/dao/EasyDAO.js:
```diff
@@ -162,13 +203,43 @@ if ( getContextualize() ) {
 
+if ( getOrder() != null &&
+     getOrder().length > 0 ) {
+  // TODO: CompositeDAO or thenBy
```
TODO

Still doesn't make much sense how it's currently written without `ThenBy` since it'll just end up using whatever comparator was last in the list.

jlhughes replied:
Yes, this definitely a TODO. It was never my intent to implement all this now.  We need to move as much of the current service setups to EasyDAO.   We also need EasyDAO to turn clustering on.

TW80000 commented on src/foam/dao/EasyDAO.js:
```diff
@@ -162,13 +203,43 @@ if ( getContextualize() ) {
+  if ( foam.nanos.auth.Authorizable.class.isAssignableFrom(getOf().getObjClass()) ) {
+    delegate = new foam.nanos.auth.AuthorizationDAO(getX(), delegate);
+  } else {
+    logger.warning("EasyDAO", "authorize=true but 'of' ",getOf().getId(), "not Authorizable");
```
In this case we're setting `authenticate` to false above but not actually adding an authorization decorator. Isn't that a security hole?

smirnovaka replied:
@jlhughes  can you please clarify. If `authorize` is set to `true` then `authenticate` goes to false automatically?

jlhughes replied:
If authorize is true and the test for instanceof passes, then we set authenticate to false.  This is temporary as we slowly transition from authenticate to authorize. Similar with read-only, we can't just set it until we go through every EasyDao setup.

TW80000 commented on src/foam/dao/EasyDAO.js:
```diff
@@ -227,6 +303,12 @@ return delegate;
       value: false
     },
+    {
+      /** Enable standard authorization. */
```
This isn't technically true with the way it's currently set up in this PR. The default `Authorizer` is `AuthorizableAuthorizer`. I don't think we have `StandardAuthorizer` yet, but @penzital is working on it if I'm not mistaken.

smirnovaka replied:
Yes, currently it will be only authorizing objects that implement authorizable but the StandardAuthorizer is in progress so I think we can keep it this way. Troy is correct(almost) @noodlemoodle is working on it.

TW80000 commented on src/tests:
```diff
@@ -43,7 +43,7 @@ p({"class":"foam.nanos.test.Test","passed":10,"failed":0,"id":"FoldAndReduceTest
 p({"class":"foam.nanos.test.Test","passed":1,"failed":0,"id":"ValidateTest","lastRun":"2019-05-07T01:03:31.645Z","lastDuration":13,"status":1,"code":"valid = new foam.nanos.auth.Phone.Builder(x).setNumber(\"123-123-1234\").build();\nvalid.validate(x);\n\ninvalid = new foam.nanos.auth.Phone.Builder(x).setNumber(\"\").build();\nthrown = false;\ntry {\n    invalid.validate(x);\n} catch (Exception e) ...
 p({ class: 'foam.nanos.auth.PreventDuplicateEmailDAOTest', id: 'PreventDuplicateEmailDAOTest' })
-p({"class":"foam.nanos.test.Test","id":"Prevent privilege escalation tests","code":"import foam.core.X;\u000aimport foam.dao.DAO;\u000aimport foam.nanos.auth.AuthorizationException;\u000aimport foam.nanos.auth.Group;\u000aimport foam.nanos.auth.GroupPermissionJunction;\u000aimport foam.nanos.auth.Permission;\u000aimport foam.nanos.auth.User;\u000aimport foam.util.Auth;\u000a\u000aimport java.util ...
+p({"class":"foam.nanos.test.Test","id":"Prevent privilege escalation tests","status":2,"code":"import foam.core.X;\nimport foam.dao.DAO;\nimport foam.nanos.auth.AuthService;\nimport foam.nanos.auth.AuthorizationException;\nimport foam.nanos.auth.Group;\nimport foam.nanos.auth.GroupPermissionJunction;\nimport foam.nanos.auth.Permission;\nimport foam.nanos.auth.User;\nimport foam.nanos.logger.Logge ...
```
@jlhughes Can you please paste the updated script as a reply to this comment so I can diff it with the current one?

jlhughes committed:
Only set Authenticate to false if AUthorize succeeds

jlhughes committed:
Merge branch 'master' of github.com:foam-framework/foam2 into CPF-2218/SecureDAOs

smirnovaka commented on src/foam/dao/EasyDAO.js:
```diff
@@ -162,13 +203,43 @@ if ( getContextualize() ) {
+     getNSpec().getServe() &&
+     ! getAuthorize() &&
+     ! getAuthenticate() &&
+     ! getReadOnly() ) {
```
why is ReadOnly important? Isn't it a common case when served service is not readOnly? Like all the daos the client is writing to?

jlhughes replied:
For 'serve'd daos, they should be authenticate, authorized, or read-only.   Nothing should be open read-write.

smirnovaka commented on src/foam/dao/EasyDAO.js:
```diff
@@ -227,6 +303,12 @@ return delegate;
+      /** Enable standard authorization. */
+      class: 'Boolean',
+      name: 'authorize',
+      value: false
```
a suggestion, maybe should be a part of second secure refactoring ... Should the default be true?

jlhughes replied:
Yes, this is coming later.  This is phase I, we can't convert everything at once.

smirnovaka commented on src/foam/nanos/notification/email/services.jrl:
```diff
@@ -1,5 +1,12 @@
-p({"class":"foam.nanos.boot.NSpec", "name":"emailTemplateDAO",                 "lazy":false,  "serve":true,  "serviceScript":"return new foam.dao.PMDAO(x, new foam.dao.PipelinePMDAO(x, new foam.dao.SequenceNumberDAO(new foam.dao.java.JDAO(x, foam.nanos.notification.email.EmailTemplate.getOwnClassInfo(), \"emailTemplates\"))));","client":"{\"of\":\"foam.nanos.notification.email.EmailTemplate\"}" } ...
+p({"class":"foam.nanos.boot.NSpec", "name":"localEmailTemplateDAO",                 "lazy":false,  "serve":false,  "serviceScript":"return new foam.dao.EasyDAO.Builder(x).setOf(foam.nanos.notification.email.EmailTemplate.getOwnClassInfo()).setAuthenticate(false).setPm(true).setPipelinePm(true).setSeqNo(true).setJournalType(foam.dao.JournalType.SINGLE_JOURNAL).setJournalName(\"emailTemplates\").bu ...
```
should it be `authenticate:false`? It's not even served

jlhughes replied:
It's EasyDAO and presently authenticate defaults to true.

smirnovaka commented on src/foam/nanos/notification/email/services.jrl:
```diff
@@ -1,5 +1,12 @@
-p({"class":"foam.nanos.boot.NSpec", "name":"emailTemplateDAO",                 "lazy":false,  "serve":true,  "serviceScript":"return new foam.dao.PMDAO(x, new foam.dao.PipelinePMDAO(x, new foam.dao.SequenceNumberDAO(new foam.dao.java.JDAO(x, foam.nanos.notification.email.EmailTemplate.getOwnClassInfo(), \"emailTemplates\"))));","client":"{\"of\":\"foam.nanos.notification.email.EmailTemplate\"}" } ...
+p({"class":"foam.nanos.boot.NSpec", "name":"localEmailTemplateDAO",                 "lazy":false,  "serve":false,  "serviceScript":"return new foam.dao.EasyDAO.Builder(x).setOf(foam.nanos.notification.email.EmailTemplate.getOwnClassInfo()).setAuthenticate(false).setPm(true).setPipelinePm(true).setSeqNo(true).setJournalType(foam.dao.JournalType.SINGLE_JOURNAL).setJournalName(\"emailTemplates\").bu ...
+
+p({"class":"foam.nanos.boot.NSpec", "name":"emailTemplateDAO",                 "lazy":false,  "serve":true,  "serviceScript":"return new foam.dao.EasyDAO.Builder(x).setOf(foam.nanos.notification.email.EmailTemplate.getOwnClassInfo()).setInnerDAO((foam.dao.DAO) x.get(\"localEmailTemplateDAO\")).build();","client":"{\"of\":\"foam.nanos.notification.email.EmailTemplate\"}" })
```
this and and pretty much all other services are set `authorize:false` (not set at all but it is the default). As I mentioned above I think the default should be true or explicitly set to true. I'd imagine only local services would not require authorization
